src/os_nix.c

```diff
@@ -69,7 +69,7 @@ static struct {
 
 static void load_bluez_symbols(void)
 {
-  void* bluetooth = dlopen("libbluetooth.so", RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE);
+  void* bluetooth = dlopen("libbluetooth.so.3", RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE);
```

Comment:
Sorry but no, this will not get merged.

Now you hardwire a specific version of the shared library, preventing the code from working on any machine that happens to have a different one? 

Before it was enough to recompile/relink the library, now we have to actually modify it? Can you imagine what sort of support nightmare this will be?

What are you trying to solve by this?

Reply:
>if the user doesn't have libbluetooth.so (.x, well spotted) on their system, I just don't show the option to connect wiimotes in my software which is an order of magnitude better than any crash

Sorry but then this is a completely wrong approach to the problem. 

The proper solution is to not hack in your application-specific thing into an upstream library but to convert the Bluetooth-dependent part of **your software** into a dynamically loaded plugin instead. Then if the user doesn't have Bluetooth, you don't load the part that requires Wiiuse. 

>well, yes, I ship appimages and the problem is that the libbluetooth.so I may ship may not be compatible with my user's system bluez version (e.g. bluez 4 vs bluez 5) - in contrast, the hci_ symbols exported from libbluetooth haven't changed for years - if i'm not mistaken, not a single symbol was changed since 2010/2011

That's not really an argument for doing this. There is zero guarantee that the ABI won't change in the future - and then what? What you are doing does not guarantee that the code won't blow up - what if Bluez 6 adds or removes an argument? Or a function is removed or renamed? Your code doesn't check for that, only extracts the functions from the loaded library and blindly calls them.

If you want to support Bluez 4 (not sure which mainstream Linux distro still ships this old version - the last Bluez 4 was released in 2012!) then you need to build multiple binaries. In such case Bluez is likely the least of your problems anyway - you will have ancient glibc, ancient kernel and many other things to deal with, we are talking 10 years old distros here. 

As I wrote above, I am not going to merge this. It is buggy and absolutely not warranted, especially given that this is a core dependency of the project in Linux.

Reply:
> What are you trying to solve by this?

the run-time issues that I had when I used to ship a libbluetooth.so it crashed when my users ran it on a different distro as the internal communication between libbluetooth.so and the other parts of the bluez stack are much less stable than the hci_lib.h API & ABI - from what I can understand you really can't mix and match two different libbluetooth.so and bluez daemons. 

It's exactly the same to me than e.g. PortAudio loading ALSA symbols through dlopen - that means someone can link against portaudio and still have their software run on a machine that does not ship libasound.so.2 (and yes, I've also seen those in the wild). But I understand that this increases the burden on wiiuse - I've been keeping this patch in my fork for a few years, it can stay a few more ; I'm confident that by the time a libbluetooth.so.4 is released with an incompatible ABI most of my users' wiimote will have broken down due to old age anyways :p 

>  then you need to build multiple binaries

well, not with this patch, that I can promise - the API & ABI of the libbluetooth functions that wiiuse uses did not change between bluez 4 and bluez 5

Reply:
I would really prefer to not merge this into Wiiuse because it addresses a rather specific concern you may have and I feel it causes more problems than it solves. 

Portaudio isn't a great analogy because that library is explicitly a wrapper for different sound systems - and tries and loads the various "backends" depending on what is available. Wiiuse isn't a wrapper with alternative backends. It cannot work without Bluetooth. So loading this dynamically would be strange. 

Concerning the runtime issue with the different libraries - have you actually considered using the Linux native support for Wiimote through the input subsystem? Wiiuse doesn't use it because it predates the Linux support by years but these days one can directly pair Wiimote to Linux, have automatic reconnection, notifications when devices are available,  connected and disconnected over DBus and standardized HID API (e.g. using the `hidapi` library). You get both the low level report interface (similar to what Wiiuse uses) and also "cooked" API where the Wiimote reports are translated to normal gamepad-like ones. 

For a Linux native application that could be a lot more powerful solution than the rather clumsy and limited historical design of Wiiuse.

Reply:
> Wiiuse doesn't use it because it predates the Linux support by years but these days one can directly pair Wiimote to Linux, have automatic reconnection, notifications when devices are available, connected and disconnected over DBus and standardized HID API (e.g. using the hidapi library). You get both the low level report interface (similar to what Wiiuse uses) and also "cooked" API where the Wiimote reports are translated to normal gamepad-like ones.

I didn't know about this, I'll investigate next time I have access to a computer with bluetooth (unless there's some way to fake it), thanks

Reply:
You could probably fake at least the HID reports by using the /dev/input/uinput device and kernel module. That allows to emulate a HID/input device. 

You want to look at xwiimote for the Wiimote support:
http://dvdhrm.github.io/xwiimote/